### DISCUSSION TITLE
Bitwise AND custom operator

### DISCUSSION THREAD

**Discussion by dalen** [alirezanet/Gridify#111] [Q&A] [answered]
I'm trying to make a custom operator for bitwise and (on top of Pomelo EF Core with MariaDB). But I can't get it working.

I was hoping something like this would work, it compiles, but doesn't work runtime

```c#
public class BitwiseAndOperator : IGridifyOperator
{
    public string GetOperator() => "#*";

    public Expression<OperatorParameter> OperatorHandler()
    {
        return (prop, value) => ((uint)prop & (uint)value) != 0;
    }
}
```

```
Rewriting child expression from type 'System.Object' to type 'AbilityLib.ComponentOperationFlags' is not allowed, because it would change the meaning of the operation. If this is intentional, override 'VisitUnary' and change it to allow this rewrite.
```

**Comment by alirezanet** [comment]
Hi @dalen,
The operator looks fine and it should work, this exception I believe is not related to Gridify, can you provide more information? I don't know what is this exception about.

**Reply by dalen** [reply]
I'm using EF core 7 with pomelo MySQL adapter. The field I'm trying with is defined as a `[Flags]` enum in the ef core db context. But on the db side it is an uint. But ideally I'd like the operator to work regardless if the backing field is uint, ushort or byte.

**Reply by alirezanet** [reply]
In this case, if you try to write the same query using LINQ, does it work? I mean it might be pomelo limitation which can't parse your expression.

**Reply by dalen** [reply]
in a simple `.Where` it works.

For example this

```
dbContext.AbilityComponents.Where(c => ((uint)c.ComponentFlags & 1) != 0)
```

I can see that being executed as this SQL
```
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (1ms) [Parameters=[@__p_1='?' (DbType = Int32), @__p_0='?' (DbType = Int32)], CommandType='Text', CommandTimeout='30']
      SELECT `m`.`Id`, `m`.`A00`, `m`.`AttackType`, `m`.`Comment`, `m`.`ComponentFlags`, `m`.`Values`
      FROM `ability_components` AS `m`
      WHERE (CAST(`m`.`ComponentFlags` AS unsigned) & 1) <> 0
      ORDER BY `m`.`Id`
      LIMIT @__p_1 OFFSET @__p_0
```

the `CAST` isn't optimal I suppose, but works

**Reply by dalen** [reply]
One workaround I've been able to make to be able to filter on flags is this

```
        return new GridifyMapper<AbilityLib.AbilityComponent>(true)
            .AddMap("componentFlags_hasflags", (c, index) => ((int)c.ComponentFlags & index) != 0);
```

then I can filter using `&filter=componentFlags_hasflag[1]=true`

Not quite as nice as a custom operator, but it does work.

**Reply by alirezanet** [reply]
I managed to reproduce the error, I'll try to fix it and hopefully, you can use the operator in the next versions...
#112

**Comment by alirezanet** [comment, marked as answer]
Hi @dalen,
Please try `v2.9.0`, custom operators should now support the unary expressions. and let me know if you had any other problems.
